src/main.rs

```diff
@@ -1,100 +1,170 @@
+use anyhow::Result;
 use aptos_crypto::ed25519::{Ed25519PrivateKey, Ed25519PublicKey};
 use aptos_keygen::KeyGen;
 use aptos_types::{
     account_address::{create_multisig_account_address, AccountAddress},
     transaction::authenticator::AuthenticationKey,
 };
-use std::env;
-use std::thread;
-use std::time::Duration;
+use chrono::prelude::{DateTime, Local};
+use clap::Parser;
+use crossbeam::channel;
+use std::thread::{self, available_parallelism};
+use std::time::{Instant, SystemTime};
 
-const PREFIX_ARG: usize = 1;
-const N_THREADS_ARG: usize = 2;
-const MULTISIG_ARG: usize = 3;
-const MULTISIG_FLAG: &str = "-m";
-const SEQUENCE_NUMBER: u64 = 0;
+/// Multisig account generation is assumed to take place in first transaction of standard account.
+const SEQUENCE_NUMBER_MULTISIG: u64 = 0;
+
+/// Optivanity: hyper-parallelized vanity address generator for the Aptos blockchain, brought to you
+/// by Econia Labs
+#[derive(Parser, Debug)]
+struct CliArgs {
+    /// Address prefix to match (no leading 0x). Each additional character slows search by 16x.
+    #[arg(short, long)]
+    prefix: String,
+    /// Use this flag if you want to search for multisig address(es)
+    #[arg(short, long)]
+    multisig: bool,
+    /// Number of vanity accounts to generate
+    #[arg(short, long, default_value_t = 1)]
+    count: u64,
+    /// Number of threads to use. Only specify if you want to use fewer cores than available.
+    #[arg(short, long, default_value_t = available_parallelism().unwrap().get())]
+    threads: usize,
+}
 
 /// Derive authentication key bytes vector from a reference to a private key.
 fn auth_key_bytes_vec(private_key_ref: &Ed25519PrivateKey) -> Vec<u8> {
     AuthenticationKey::ed25519(&Ed25519PublicKey::from(private_key_ref)).to_vec()
 }
 
-fn main() {
-    // Get command arguments.
-    let args: Vec<_> = env::args().collect();
-    let mut check_str = args[PREFIX_ARG].clone(); // Get prefix string for format check.
-    let is_odd_length = check_str.len() % 2 != 0; // Check if string has odd character count.
-    if is_odd_length {
-        // Append a 0 to check string if an odd number of hex characters.
-        check_str.push('0')
+/// Parse command line arguments, verifying hex characters and specified thread count.
+fn parse_args() -> Result<CliArgs> {
+    let args = CliArgs::parse();
+
+    // Verify prefix has valid hex characters, appending a 0 as needed for an odd character count.
```

Comment:
Isn't this not true? If I provide `aaaaa` then the character in the sixth position can be anything, not just zero. You're mutating the user's description of what they want in this case I think (not good). Will slow down search by 16x for odd prefixes!

Edit: I see you _try_ to pop off this character further down. Why is it added in the first place?

Reply:
The character is added here to check if the rest of the prefix is valid hex.

The `decode` function only accepts even-numbered strings because it takes two hex characters to specify a byte.